py/objset.c

```diff
@@ -227,47 +227,47 @@ static mp_obj_t set_diff_update(size_t n_args, const mp_obj_t *args) {
 }
 static MP_DEFINE_CONST_FUN_OBJ_VAR(set_diff_update_obj, 1, set_diff_update);
 
-static mp_obj_t set_intersect_int(mp_obj_t self_in, mp_obj_t other, bool update) {
+static mp_obj_t set_intersect_int(size_t n_args, const mp_obj_t *args, bool update) {
+    mp_obj_set_t *in_ptr;
     if (update) {
-        check_set(self_in);
+        check_set(args[0]);
+        in_ptr = MP_OBJ_TO_PTR(args[0]);
     } else {
-        check_set_or_frozenset(self_in);
+        check_set_or_frozenset(args[0]);
+        in_ptr = MP_OBJ_TO_PTR(set_copy(args[0]));
     }
 
-    if (self_in == other) {
-        return update ? mp_const_none : set_copy(self_in);
-    }
+    for (size_t i = 1; i < n_args; i++) {
+        if (MP_OBJ_FROM_PTR(in_ptr) != args[i]) {
+            mp_obj_set_t *out_ptr = MP_OBJ_TO_PTR(mp_obj_new_set(0, NULL));
 
-    mp_obj_set_t *self = MP_OBJ_TO_PTR(self_in);
-    mp_obj_set_t *out = MP_OBJ_TO_PTR(mp_obj_new_set(0, NULL));
+            mp_obj_t iter = mp_getiter(args[i], NULL);
+            mp_obj_t next;
+            while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
+                if (mp_set_lookup(&in_ptr->set, next, MP_MAP_LOOKUP)) {
+                    set_add(MP_OBJ_FROM_PTR(out_ptr), next);
+                }
+            }
 
-    mp_obj_t iter = mp_getiter(other, NULL);
-    mp_obj_t next;
-    while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
-        if (mp_set_lookup(&self->set, next, MP_MAP_LOOKUP)) {
-            set_add(MP_OBJ_FROM_PTR(out), next);
+            m_del(mp_obj_t, in_ptr->set.table, in_ptr->set.alloc);
```

Comment:
Do I understand correctly that in the `!update` case, the behaviour is:

1. Make a full copy of set `self`.
2. For each argument, allocate a new set and build it up to have the items in both `self` and the argument.
3. Delete `self`'s set data, and then swap in the new set.

This seems like a lot of temporary memory churn, especially in the simplest case where there's only one argument.

I know when you first asked me about this on Discord I said it'd be better to build up the new set by adding instead of copying it and then removing items, to limit the peak memory usage. However if this is the alternative then maybe it is better to copy the set and then remove items?

The only downsize I can see is that the final result table might be too big (i.e. started with many entries and then removed most). Currently `mp_set_rehash` (which re-allocates and rebuilds the map) only gets called on growth never when making the set smaller. Could still emulate this though, by checking at the end if the set has a lot of free space and copying it if it does...?

Reply:
> This seems like a lot of temporary memory churn, especially in the simplest case where there's only one argument.

If this is the only problem, I can solve it easily without significant code size increase.

About the idea of creating a set and delete items, we should do some linear searches because we want to make sure all items of set `X` are present in `Iterable[i]` and If it's not, then we remove it from `X`. It would be much easier if all of `args` were sets.
IMHO we should choose between these options:
1. Focus on reducing memory footprints with cost of decreasing performance (Linear search)
2. Focus on increasing performance with cost of increasing temporary memory usages.

Reply:
If there's a simple way to avoid the unncessary copy in the single arg version, then that sounds good.

I don't understand how copying the set once and then removing items requires linear searches though. What about something like this:

```c
p_obj_set_t *out_ptr = MP_OBJ_TO_PTR(set_copy(self_in));
for (size_t i = 1; i < n_args; i++) {
        if (MP_OBJ_FROM_PTR(in_ptr) != args[i]) {
            mp_obj_t iter = mp_getiter(args[i], NULL);
            mp_obj_t next;
            while ((next = mp_iternext(iter)) != MP_OBJ_STOP_ITERATION) {
                 if (!mp_set_lookup(&in_ptr->set, next, MP_MAP_LOOKUP)) {
                    set_discard(MP_OBJ_FROM_PTR(out_ptr), next);
                }
       }
}
```

This is the same structure as the current PR, only with the inner `set_add` swapped to a `set_discard`. Because any item stays removed once you discard it, you only have to copy the set once no matter how many args are provided.

You have thought about this problem a lot more than me, though. I am probably missing an important detail.

Reply:
Let me show you what is wrong with your solution with an example: If we call `set.intesection({1,2},{2})` then:

1. You're making a copy of `{1,2}.`
2. Then iterating over `{2}` and checking 2 is present in `{1,2}` (which is present, so you do nothing)
3. Finally, the result will be `{1,2}` which is not correct.

As I told you earlier, we should check all items of `self_in` are present in args[i] and remove those are not. That's why this approach leads us to linear search.

Reply:
:facepalm: Makes sense, thanks for the clear explanation.